### ISSUE TITLE
Add support for pattern load testing scenarios

### ISSUE THREAD

**Issue by catalin-me** [mcollina/autocannon#235]
## Description

This issue proposes adding advanced testing scenario features to Autocannon that include the possibility of running configurable load tests that support ramp-up, spike and cooldown behaviours.

## Proposed Solution

This proposal introduces 3 new configuration properties that would allow configuring such behaviour:

- `rampUp`
- `spike`
- `coolDown`

Each new configuration property would also support the following properties that would model the pattern behaviour appropriately:

- `connections` The amount of connections.
- `fixed` The amount of time the phase should last. Contextual if `connections` is defined.
- `increase` The amount of connections to increase on the defined `every` interval.  Contextual if  `every` is used.
- `every` The interval of time in seconds that should pass before the `connections` are increased. Contextual if `increase` is used.

The following would demonstrate a sample configuration object:

    const options = {
      url: "<URL>",
      connections: 10,
      duration: 120,
      pattern: {
        rampUp: {
          connections: 100,
          fixed: 30,
          increase: 100,
          every: 10
        },
        spike: {
          connections: 500,
          fixed: 30,
          increase: 100,
          every: 10
        },
        coolDown: {
          connections: 100,
          fixed: 30,
          increase: 100,
          every: 10
        }
      }
    }

### Key Considerations

- The base `connections` property would be ignored if a pattern config is provided as each stage would define its own `connections` target.
- The base `duration` property would be used to determine how long it would take to complete the entire load testing pattern. If ignored, the `fixed` value would be used instead to determine the total duration of the pattern test.
- Controlling the base `duration` property would allow to create repetitive patterns if required.

### Caveats

- The only configurable property would be the `connections` property for pattern load tests. Therefore, developers would be constrained by the existent provided values in the primary configuration object as HTTP clients would not be able to be bootstrapped from scratch between pattern phases due to performance limitations.

### Notes

@mcollina I've already worked on a POC for this proposal. Please let me know your thoughts and suggestions and I will be more than happy to provide a PR.

**Comment by mcollina** [comment]
Could you please clarify what "spike" behavior would be?

Feel free to send the PR anyway.

**Comment by Kefir100** [comment]
Hey, awesome library, but this feature is really missing)

**Comment by Kefir100** [comment]
But I think there should be some different interface, something like this
```javascript
[
    {
      /**
       * This one ramps up to 50 connections from initial 10 (default) during 5 secs
       * That means it increases connections on (50-10)/5 = 8 connections/sec
       */
        title: 'ramp-to-50-for-5-secs',
        duration: 5,
        target: 50,
    },
    {
      /**
       * This one cools down to 5 connections from previous 50 connections during 4 secs
       * That means it decreases connections on (10-50)/4 = -10 connections/sec
       */
        title: 'cool-down-to-10-for-10-secs',
        duration: 4,
        target: 10,
    }
]
```
The formula for stage's connection/sec delta is `(target-currentConnectionValue)/duration`

**Comment by mcollina** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by Kefir100** [comment]
Yes, I would like, but ,unfortunately, I cannot promise. I ll send a PR as soon(/if) I manage to create one)
